### PR TITLE
Vb 1358 extend time to allow cancellation of visits after visit start time

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Extend the time till when visits can be cancelled via a configurable parameter.

## What is the intent behind these changes?

To allow expired visits to be cancelled as NOMIS allows expired visits to be cancelled.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt [hunk 1]
```diff
@@ -263,8 +265,12 @@ class VisitService(
   fun cancelVisit(reference: String, cancelOutcome: OutcomeDto): VisitDto {
     val visitEntity = visitRepository.findBookedVisit(reference) ?: throw VisitNotFoundException("Visit $reference not found")
 
+    var visitCancellationDateAllowed = LocalDateTime.now()
     // check if the visit being cancelled is in the past
-    validateVisitStartDate(visitEntity, "cancelled")
+    if (visitCancellationDayLimit > 0) {
+      visitCancellationDateAllowed = visitCancellationDateAllowed.minusDays(visitCancellationDayLimit.toLong()).truncatedTo(ChronoUnit.DAYS)
+    }
+    validateVisitStartDate(visitEntity, "cancelled", visitCancellationDateAllowed)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:273]: Would move this into a method.. to encapsulate the complexity and make code easy to read
`    val visitCancellationDateAllowed = getAllowedCancellatonDate()`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt [hunk 2]
```diff
@@ -334,8 +340,12 @@ class VisitService(
     )
   }
 
-  private fun validateVisitStartDate(visit: Visit, action: String) {
-    if (visit.visitStart.isBefore(LocalDateTime.now())) {
+  private fun validateVisitStartDate(
+    visit: Visit,
+    action: String,
+    allowedVisitStartDate: LocalDateTime = LocalDateTime.now()
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:346]: why the default?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/legacy/CancelVisitTest.kt [hunk 3]
```diff
@@ -281,13 +290,230 @@ class CancelVisitTest(@Autowired private val objectMapper: ObjectMapper) : Integ
...
+
+    // Then
+    responseSpec.expectStatus().isOk
+      .expectBody()
+      .jsonPath("$.visitStatus").isEqualTo(VisitStatus.CANCELLED.name)
+      .jsonPath("$.outcomeStatus").isEqualTo(OutcomeStatus.CANCELLATION.name)
+      .returnResult()
+
+    // And
+    verify(telemetryClient, times(1)).trackEvent(eq("visit-cancelled"), any(), isNull())
```
Comment (AledWEvans) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/legacy/CancelVisitTest.kt:322]: not sure you have to test for events as other test do that,... alll your doing here is testing that you can cancel.. so status is enough

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/visit/CancelVisitTest.kt [hunk 4]
```diff
@@ -280,4 +290,197 @@ class CancelVisitTest(@Autowired private val objectMapper: ObjectMapper) : Integ
...
+    val expiredVisit = visitEntityHelper.create(visitStatus = BOOKED, visitStart = visitStart)
+
+    // When
+    val responseSpec = callCancelVisit(webTestClient, setAuthorisation(roles = listOf("ROLE_VISIT_SCHEDULER")), expiredVisit.reference, outcomeDto)
+
+    // Then
+    responseSpec.expectStatus().isOk
+      .expectBody()
+      .jsonPath("$.visitStatus").isEqualTo(VisitStatus.CANCELLED.name)
+      .jsonPath("$.outcomeStatus").isEqualTo(OutcomeStatus.CANCELLATION.name)
```
Comment (AledWEvans) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/visit/CancelVisitTest.kt:314]: only have to check for cancelled status...

as other test cover out come!
"cancel visit by reference -  with outcome and outcome text"

Or you could have a common method something like 
assertVisitCancellation() and in that you could check all asserts e.g.

 status
 outcome 
 events

and reuse across test class

**Review by AledWEvans — CHANGES_REQUESTED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:273]: Agreed, will update.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/legacy/CancelVisitTest.kt (hunk 3 above)
Comment (dhirajshettymoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/legacy/CancelVisitTest.kt:322]: Agreed. I will refactor to use the assertVisitCancellation method as suggested below and remove some asserts that are not needed.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt:271]: Updated to add getAllowedCancellatonDate method.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/visit/CancelVisitTest.kt (hunk 4 above)
Comment (dhirajshettymoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/visit/CancelVisitTest.kt]: Updated based on the above suggestions. thanks.

**Review by AledWEvans — APPROVED** [review]
LGTM

**Pull request merged**